### ISSUE TITLE
Electron: initialise crash reporter before app.on("ready")

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#97123]
Steps to Reproduce (get latest from `master`):

1. add a `process.crash()` somewhere into workbench code that executes after a timeout
2. run `scripts/code.sh --crash-reporter-directory <path to empty folder>`
3. wait for the crash
4. check the output of the crash folder

=> no `dmp` file exists (but other files that seem to indicate the crash directory is picked up)

This works fine when triggering a crash on the `main` side. So I wonder if we are doing something wrong with how we setup the crash reporter.

**Comment by bpasero** [comment]
I actually see the same for forked process.

**Comment by bpasero** [comment]
Here is the result of my OS testing:
* Windows
  * `main`: success
  * `renderer`: success
  * `child-process`: success
* macOS
  * `main`: success
  * `renderer`: fail
  * `child-process`: fail
* Linux
  * `main`: success
  * `renderer`: success
  * `child-process`: fail

**Comment by deepak1556** [comment]
On windows and macOS where the underlying crash handler called `Crashpad` starts monitoring crashes in child processes (this includes render process, gpu process, child node process that are attached to the parent) from the moment its initialized in the main process. So any process that is created before the crash reporter is initialized in the main process will not be monitored. Currently we do the initialization from `vs/workbench/electron-browser/window.ts` which sort of covers this , but it is best to move the main process crash reporter initialization to `src/main.js` preferably before `app.on(`ready`)`. This should fix the issue you are seeing on macOS and will also help with transition for upcoming refactor.

For Linux where the crash handler is `Breakpad` , it needs to be initialized in respective process and we are doing it correctly. The child process issue here is more of an upstream problem thats fixed in E9.

**Comment by bpasero** [comment]
Awesome, thanks for clarifying. I do not think we can change this for the normal case because the crash reporter can be disabled via setting and we do not read settings so early. But I pushed a change to start the crash reporter right in the beginning when `--crash-reporter-directory` is provided, please check:

https://github.com/microsoft/vscode/blob/004282afb048d2d6a92e380ea6dbf9d7afd809bd/src/main.js#L60-L66

**Comment by deepak1556** [comment]
Change looks good 👍 

>  I do not think we can change this for the normal case because the crash reporter can be disabled via setting and we do not read settings so early.

We have to deprecate this and only allow disabling crash reporter via `argv.json` , because with E9 crash reporter can only be initialized from the main process. If we allow users to configure via `argv.json`, we can restart the app and launch with `--disable-crash-reporter` flag and respect that.

**Comment by bpasero** [comment]
Sounds good, let me rephrase this issue.

**Comment by deepak1556** [comment]
This is done :)
